design/file-descriptors.md

```diff
@@ -0,0 +1,14 @@
+# File Descriptors
+File Descriptors ([witx type `fd`](https://github.com/WebAssembly/WASI/blob/master/phases/snapshot/docs.md#-fd)) are an abstract way of representing access to a resource. More often than not, they are processes (WIP), sockets (WIP), files, or even the console.
+
+# The Console
+When a program is initiated, `stdout`, `stdin`, and `stderr` (respectively `fd` ids `0`, `1`, and `2`) are opened. To print to these, simply call `fd_write` with the corresponding arguments. An simple "Hello World!" program written in `wat` is shown below:
+
+```wat
+(module
+  ;; TODO: small hello world?
+)
+```
```

Comment:
Wasmtime's testsuite has [a hello world test](https://github.com/bytecodealliance/wasmtime/blob/main/tests/wasm/hello_wasi_snapshot1.wat) which you can base this on.

It's admittedly a little inconvenient to set up the iovec structures. The idea of adding more convenient non-vectored `read` and `write` functions to WASI has come up a few times; I'd be interested to hear if to think that'd be worth doing.

Reply:
**TL;DR** nah

Personally, I think that less code is better from the perspective of the (Wasmtime/Wasi/Wasm/etc) library developer, and as such, there should not be more convenient methods. Here's why:

- **One way to do things.**
Consistency throughout the documentation is important, and by forcing people to go the iovec route you're ensuring things will be done correctly. Iovecs are the "correct" way to do things, because they discourage common performance pitfalls (i.e. string concatenation).
- **Smaller specification to conform to.**
By not having the overload, your interpreter does not grow in size. It stays small, and people who wish to implement the specification can do so easily and reasonably. Staying small in turn encourages widespread adoption. For example, the x86_64 standard is so huge and bloated that it's very difficult to create emulators for them (some are VirtualBox, Vmware, and QEMU) while a language such as `C` is adopted nearly universally at this point due to it's extreme simplicity (but its age also causes it to be complicated in its own rights). Admittedly, this comparison isn't very fair, but it's what I was able to think of on the top of my head.
On the contrary, for features that'd be too hard for an interpreter to get right, it should be implemented into the instruction set - such as vectorization. It just makes more sense for the tool generating the Wasm to generate code that can vectorize appropriately, and the cost of trying to write an interpreter to correctly take advantage of vectorization patterns would just lead to more bloat then letting the compiler do it.
- **Wasm has no need to optimize for the human.**
Just like how you'd never write assembly unless you're starving for those 0.27 nanoseconds, one should never have to write web assembly either. It's primarily generated by other compilers which target web assembly and whose generated outputs will get better as time progresses on. There's simply no need to optimize for the human here.
- **Encourages performant code.**
This is highly subjective, but what I imagine is that given an iovec is setup to use in the first place, it can most likely be reused in future operations. An overload for `read`/`write` would most likely (take this with a grain of salt because i only barely have an idea of what i'm talking about lol) have to setup its own iovec, copy the input into it, and then call `file_write` (or vice versa - setting up an iovec and marshalling the resulting bytes as whatever structure).
- **Having less allows more to be done.**
Instead of having an extra method or two and having to care for it and possibly any number of others that are introduced too as little helpers (documentation, ensuring correct code, deprecation of features, etc.), you can primarily focus the engineering effort that would be spent into maintaining that (as well as the other numerous projects that'd have to cope with the overload) into optimizing more pressing concerns - such as introducing a way to have no-copy writes (i.e. string into iovec being free, or some other datastructure which gets the job done in a similar fashion)

Apologies if any point is unclear or what have you, I feel like I rambled for far too long - hopefully that helps you make a decision. Of course, I'm not any designer at Wasm so I don't have a clear picture for the goals of the project or how it all works together as a whole, so read my comments with not one but two grains of salt.

Oh, and I added hello world example you mentioned.